pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,967 @@
+								invest_amount,
+							},
+						)
+					}
+				}
+			}
+			// Reduce amount of foreign by the increasing amount and increase investing as well as
+			// foreign_done amount by the minimum of active swap amounts
+			Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+				swap: foreign_swap,
+				done_amount,
+			} => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let invest_amount = swap.amount.min(foreign_swap.amount);
+				let done_amount = invest_amount.ensure_add(*done_amount)?;
+
+				// pool swap amount is immediately invested and done amount increased equally
+				match swap.amount.cmp(&foreign_swap.amount) {
+					Ordering::Equal => Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+						done_swap: Swap {
+							amount: done_amount,
+							..*foreign_swap
+						},
+						invest_amount,
+					}),
+					// swap amount is immediately invested and done amount increased equally
+					Ordering::Less => {
+						Ok(
+							Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+								swap: Swap {
+									// safe since swap.amount < foreign_swap.amount
+									amount: foreign_swap.amount - swap.amount,
+									..*foreign_swap
+								},
+								done_amount,
+								invest_amount,
+							},
+						)
+					}
+					// foreign swap amount is immediately invested and done amount increased equally
+					Ordering::Greater => {
+						Ok(
+							Self::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+								swap: Swap {
+									// safe since swap.amount > foreign_swap.amount
+									amount: swap.amount - foreign_swap.amount,
+									..swap
+								},
+								done_amount,
+								invest_amount,
+							},
+						)
+					}
+				}
+			}
+			// Reduce amount of foreign swap by increasing amount and increase investing as well as
+			// foreign_done amount by minimum of swap amounts
+			Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: foreign_swap,
+				done_amount,
+				invest_amount,
+			} => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let invest_amount =
+					invest_amount.ensure_add(swap.amount.min(foreign_swap.amount))?;
+				let done_amount = swap
+					.amount
+					.min(foreign_swap.amount)
+					.ensure_add(*done_amount)?;
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					// pool swap amount is immediately invested and done amount increased equally
+					Ordering::Equal => Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+						done_swap: Swap {
+							amount: done_amount,
+							..*foreign_swap
+						},
+						invest_amount,
+					}),
+					// swap amount is immediately invested and done amount increased equally
+					Ordering::Less => Ok(
+						Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+							swap: Swap {
+								// safe since swap.amount < foreign_swap.amount
+								amount: foreign_swap.amount - swap.amount,
```

Comment:
I would make all these substractions safe with `ensure_`. It's true that all of them can be "ensured" from the same method, but I think these methods are quite long, and it is easy to make a fix/change in the amounts of computation that breaks the commented invariants.